src/main/java/iudx/aaa/server/registration/RegistrationServiceImpl.java

```diff
@@ -615,6 +576,211 @@ public RegistrationService getUserDetails(List<String> userIds,
+        promise.fail(new ComposeException(r));
+        return;
+      }
+      checkOrgRequired = pool.withConnection(
+          conn -> conn.preparedQuery(SQL_FIND_ORG_BY_ID).execute(Tuple.of(orgId.toString())).map(
+              rows -> rows.iterator().hasNext() ? rows.iterator().next().getString("url") : null));
+    } else {
+      checkOrgRequired = Future.succeededFuture(NO_ORG_CHECK);
+    }
+
+    Future<Void> validateOrg = CompositeFuture.all(checkOrgRequired, email).compose(x -> {
+      String url = (String) x.list().get(0);
+      String emailId = (String) x.list().get(1);
+
+      String emailDomain = emailId.split("@")[1];
+
+      if (emailId.length() == 0) {
+        Response r = new ResponseBuilder().status(400).type(URN_INVALID_INPUT)
+            .title(ERR_TITLE_USER_NOT_KC).detail(ERR_DETAIL_USER_NOT_KC).build();
+        return Future.failedFuture(new ComposeException(r));
+      }
+
+      if (url == null) {
+        Response r = new ResponseBuilder().status(400).type(URN_INVALID_INPUT)
+            .title(ERR_TITLE_ORG_NO_EXIST).detail(ERR_DETAIL_ORG_NO_EXIST).build();
+        return Future.failedFuture(new ComposeException(r));
+
+      } else if (!url.equals(emailDomain) && !url.equals(NO_ORG_CHECK)) {
+        Response r = new ResponseBuilder().status(400).type(URN_INVALID_INPUT)
+            .title(ERR_TITLE_ORG_NO_MATCH).detail(ERR_DETAIL_ORG_NO_MATCH).build();
+        return Future.failedFuture(new ComposeException(r));
+      }
+
+      return Future.succeededFuture();
+    });
+
+    List<Roles> rolesForKc =
+        requestedRoles.stream().filter(x -> x != Roles.PROVIDER).collect(Collectors.toList());
+
+    List<Tuple> roleDetails = roles.entrySet().stream()
+        .map(p -> Tuple.of(user.getUserId(), p.getKey().name(), p.getValue().name()))
+        .collect(Collectors.toList());
+
+    /* supplier to create tuple for org update */
+    Supplier<Tuple> updateOrgIdTup = () -> {
+      if (checkOrgRequired.result() == NO_ORG_CHECK) {
+        return Tuple.of(null, user.getUserId());
+      }
+      return Tuple.of(request.getOrgId(), user.getUserId());
+    };
+
+    validateOrg
+        .compose(res -> pool
+            .withTransaction(conn -> conn.preparedQuery(SQL_CREATE_ROLE).executeBatch(roleDetails)
+                .compose(
+                    success -> conn.preparedQuery(SQL_UPDATE_ORG_ID).execute(updateOrgIdTup.get()))
+                .compose(success -> kc.modifyRoles(user.getKeycloakId(), rolesForKc))))
+        .onSuccess(success -> {
+          List<Roles> updatedRoles = user.getRoles();
+          updatedRoles.addAll(rolesForKc);
+
+          JsonObject resp = new JsonObject().put("roles", updatedRoles);
+          promise.complete(resp);
+        }).onFailure(e -> {
+          if (e instanceof ComposeException) {
+            promise.fail(e);
+            return;
+          }
+
+          LOGGER.error(e.getMessage());
+          promise.fail("Internal error");
+        });
+    return;
```

Comment:
Since there are compose failures in the method itself, we fail the promise and pass on the failure to the service